runtime/domain.c

```diff
@@ -1482,29 +1490,92 @@ void caml_reset_young_limit(caml_domain_state * dom_st)
   }
 }
 
+Caml_inline void advance_global_major_slice_epoch (caml_domain_state* d)
+{
+  uintnat old_value;
+
+  CAMLassert (atomic_load (&caml_major_slice_epoch) <=
+              atomic_load (&caml_minor_collections_count));
+
+  old_value = atomic_exchange (&caml_major_slice_epoch,
+                               atomic_load (&caml_minor_collections_count));
+
+  if (old_value != atomic_load (&caml_minor_collections_count)) {
+    /* This domain is the first one to use up half of its minor heap arena
+        in this minor cycle. Trigger major slice on other domains. */
+    if (caml_plat_try_lock(&all_domains_lock)) {
+      /* Note that this interrupt is best-effort. If we get the lock,
+          then interrupt all the domains. If not, either some other domain
+          is calling for a stop-the-world section interrupting all the
+          domains, or a domain is being created or terminated. All of these
+          actions also try to lock [all_domains_lock] mutex, and the above
+          lock acquisition may fail.
+
+          If we don't get the lock, we don't interrupt other domains. This
+          is acceptable since it does not affect safety but only liveness --
+          the speed of the major gc. The other domains may themselves fill
+          half of their minor heap triggering a major slice, or do it right
+          after their next minor GC when they observe that their
+          domain-local [Caml_state->major_slice_epoch] is less than the
+          global one [caml_major_slice_epoch]. */
+      for(int i = 0; i < stw_domains.participating_domains; i++) {
+        dom_internal * di = stw_domains.domains[i];
+        stw_request.participating[i] = di->state;
```

Comment:
I don't think you should set this - it isn't needed and also it might clobber `stw_request` which isn't entirely protected by `all_domains_lock` in `caml_try_run_on_all_domains_with_spin_work`. 

Remove this line?

Reply:
Good catch. Looks like a copy-paste error. I'll remove this line. 

I'll also remove the assertion on the next line:

```c
CAMLassert(!d->interruptor.interrupt_pending);
```

It is unclear to me why this is expected to hold.